jax/_src/array.py

```diff
@@ -700,54 +700,81 @@ def make_array_from_callback(
 
   Args:
-    shape : Shape of the ``jax.Array``.
-    sharding: A ``Sharding`` instance which describes how the ``jax.Array`` is
-      laid out across devices.
-    arrays: Sequence of ``jax.Array``\s that are on a single device.
+    shape : Shape of the output ``jax.Array``. This conveys information already included with
+      ``sharding`` and ``arrays`` and serves as a double check.
+    sharding: Sharding: A global Sharding instance which describes how the output jax.Array is laid out across devices.
+    arrays: Sequence of ``jax.Array``\s that are each single device addressable. ``len(arrays)``
+      must equal ``len(jax.local_devices())`` and the shape of each array must be the same. For multiprocess code,
+      the each process will call with a different ``arrays`` argument that corresponds to that processes' data.
 
   Returns:
-    A ``jax.Array`` from a sequence of ``jax.Array``\s on a single device.
+    A global ``jax.Array``, sharded as ``sharding``, with shape equal to ``shape``, and with per device
+      contents matching ``arrays``.
 
-  Example:
+  Examples:
+
+    In this single-process example, we use ``make_array_from_single_device_arrays`` to create an
+    a global array.
 
     >>> import math
     >>> from jax.sharding import Mesh
     >>> from jax.sharding import PartitionSpec as P
     >>> import numpy as np
     ...
+    >>> mesh_rows = 2
+    >>> mesh_cols =  jax.device_count() // 2
+    ...
     >>> global_shape = (8, 8)
-    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(2, 4), ('x', 'y'))
-    >>> sharding = jax.sharding.NamedSharding(global_mesh, P('x', 'y'))
+    >>> mesh = Mesh(np.array(jax.devices()).reshape(mesh_rows, mesh_cols), ('x', 'y'))
+    >>> sharding = jax.sharding.NamedSharding(mesh, P('x', 'y'))
     >>> inp_data = np.arange(math.prod(global_shape)).reshape(global_shape)
     ...
     >>> arrays = [
-    ...     jax.device_put(inp_data[index], d)
-    ...     for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
+    ...    jax.device_put(inp_data[index], d)
+    ...        for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
     ...
     >>> arr = jax.make_array_from_single_device_arrays(global_shape, sharding, arrays)
-    >>> arr.addressable_data(0).shape
-    (4, 2)
-
-    In multi-process case, if the input is process local and data parallel
-    i.e. each process receives a different part of the data, then you can use
-    `make_array_from_single_device_arrays` to create a global jax.Array
+    >>> assert arr.shape == (8,8) # arr.shape is (8,8) regardless of jax.device_count()
+
+    When using multiple processes, the most common case is the input is process local and data parallel.
+    (Equivalently, we assume each process receives a different part of the data and the
+    total number of rows is at least one per device.) In this case, the following recipe will use
```

Comment:
```suggestion
    When using multiple processes, a common scenario is data parallelism across devices,
    with each device receiving at least one example. In this case, the following recipe will use
```
Something like this? I found the current language confusing